### DISCUSSION TITLE
[Feature Request]: More extensions

### DISCUSSION THREAD

**Discussion by nam-co** [beyondcode/herd-community#184] [Ideas]
### Feature Description

Instead of opening an issue for each extension we wish, easier like this

### Please don't put issues or duplicates (asking for an extension that is already on the list) just thumbs up

**Comment by nam-co** [comment]
FFMPEG: Like imagick this compresses videos

**Comment by nam-co** [comment]
puppeteer: Useful for packages like browsershot (amazing package)

**Reply by jbraband** [reply]
Is there a path to using the browsershot >puppeteer > headless chrome in Laravel Herd in 2025?

**Comment by rafaelgrube** [comment]
I often use the oci8 extension. It would be awesome if I can add other extensions.

**Reply by guillaumemolter** [reply]
@rafaelgrube I'm curious if you have ever been able to use/configure oci8 on Herd? Our team is currently using Homestead with a custom script to install and configure oci8, and oci8 is essentially the last blocker to us migrating to Herd.

**Comment by pyrou** [comment]
Would be definitely nice to let us pick which extension we would like to enable.

At least provide a per-php-version php.ini file, so advanced user can add their own extension=/absolute/path.so, because so far, since the php.ini file is shared across all version, no chance to add an extension that would work for multiple php version.

**Comment by ryanmortier** [comment]
`sqlsrv`
`pdo_sqlsrv`

SQL Server is a database that has first-party support through Laravel.

**Reply by KiritoXD01** [reply]
@mpociot is there any docs to follow about this one?

**Reply by ryanmortier** [reply]
@KiritoXD01 it's now included by default. https://herd.laravel.com/docs/macos/technology/php-extensions#sqlsrv

Make sure you install the ODBC drivers: https://herd.laravel.com/docs/macos/troubleshooting/odbc-drivers

**Reply by KiritoXD01** [reply]
@ryanmortier it does, but looks like is not included in the Windows Version

**Comment by renedekat** [comment]
xdebug

A debugger is an essential tool during development.

**Comment by grantholle** [comment]
ldap

**Comment by mpociot** [comment]
Taken from #47  and #26 
* GRPC
* memcached

Right now, we'll just gather feedback for additional extensions and then see if/which ones we can bundle into Herd.

**Reply by piperone** [reply]
Would be awesome to see Memcached as a service one can add from the services-menu as well.

**Comment by bradgenereux** [comment]
Ghostscript: Imagick requires it for PDF operations.

**Comment by eliurkis** [comment]
> `sqlsrv` `pdo_sqlsrv`
> 
> SQL Server is a database that has first-party support through Laravel.

🙏

**Comment by ryanscherler** [comment]
It seems that `mbregex` isn't explicitly compiled with the included `mbstring` extension and causing issues with running Craft CMS via Herd (had to revert back to Homebrew 😔). Anybody else experience the same issue?

The error I get when running any version of Herd PHP:

```php
error \mb_regex_encoding('UTF-8')

// call to undefined function mb_regex_encoding()
```

**Comment by mpociot** [comment]
> It seems that `mbregex` isn't explicitly compiled with the included `mbstring` extension and causing issues with running Craft CMS via Herd (had to revert back to Homebrew 😔). Anybody else experience the same issue?
> 
> 
> 
> The error I get when running any version of Herd PHP:
> 
> 
> 
> ```php
> 
> error \mb_regex_encoding('UTF-8')
> 
> 
> 
> // call to undefined function mb_regex_encoding()
> 
> ```

Please make sure that you've updated the built in PHP version via the settings screen.

**Comment by ryanscherler** [comment]
Oh ok - has this extension been enabled and addresses this error?

**Comment by Plytas** [comment]
> Taken from #47  and #26 
> * GRPC
> * memcached
> 
> Right now, we'll just gather feedback for additional extensions and then see if/which ones we can bundle into Herd.

Protobuf too.

**Comment by vincent-tarrit** [comment]
Hello ! Thank you for the app ! It looks really nice ! I'm just waiting to download because I would need the extension 

IMAP

To be sure to use in all my project 😃 

Thank you and I hope it will be added 
Vincent

**Comment by ewilan-riviere** [comment]
I love Herd! 😄

What about pcov for tests coverage?

**Comment by mpociot** [comment]
@ewilan-riviere I'm currently working on compiling xDebug for all PHP versions so that we can ship it with Herd, which you can then use for test coverage 👍

**Comment by alcaeus** [comment]
@mpociot is there any progress on the MongoDB support? Anything we can help with in that regard?

**Comment by pryley** [comment]
> @ewilan-riviere I'm currently working on compiling xDebug for all PHP versions so that we can ship it with Herd, which you can then use for test coverage 👍

[pcov](https://github.com/krakjoe/pcov) is almost 5 times faster than xdebug for PHPUnit code coverage.

> The differences in performance of Xdebug and PCOV are not slight. Xdebug is first and foremost a debugging extension, and when you load it, you incur the overhead of a debugger even when it's disabled. PCOV is less than 1000 lines of code (not including CFG) and doesn't have anything like the overhead of a debugger.

**Comment by BurakBoz** [comment]
gd extension doesn't have avif support yet is this too hard build statically?

**Comment by esseremmerik** [comment]
We use php_xsl extension, would love to see this included

**Comment by Plytas** [comment]
I believe the best approach here would be to provide tooling to build our own binaries. A default binary with most popular extensions is always good, but people will constantly ask for more.

**Comment by mpociot** [comment]
@Plytas the main problem is that it's not as easy to compile extensions statically depending on the extension dependencies. 
Some might be super easy to add, while others are more complicated

**Comment by Plytas** [comment]
@mpociot would you mind sharing more in depth about what it takes to build a binary with required extensions? Maybe someone in the community will be interested in the challenge to solve this?

**Comment by ianlandsman** [comment]
A second vote for the IMAP extension if it's possible.

**Comment by naingyy** [comment]
Is there any workaround to enable the **oci8** extension for now?

**Comment by jacobshenning** [comment]
@mpociot Have you looked into compiling openswoole at all? It would be awesome if that was a possibility.

**Comment by ndberg** [comment]
php-decimal

Correctly-rounded, arbitrary precision decimal floating-point arithmetic in PHP 

[https://php-decimal.io/#introduction](https://php-decimal.io/#introduction)
[https://github.com/php-decimal/ext-decimal](https://github.com/php-decimal/ext-decimal)

**Comment by morrislaptop** [comment]
'swoole'

Swoole provides concurrency for PHP with Laravel Octane.

From #97

**Comment by jacobshenning** [comment]
Have any decisions been made to not implement an extension? If so, is there a way I can see the won't implement list?

**Comment by vikas5914** [comment]
@pryley You can download the Pre-complied extension from here: https://packages.macports.org/php82-pcov/

@morrislaptop  Swoole: https://packages.macports.org/php82-swoole/

**Comment by stojankukrika** [comment]
Mongo db will be nice to have as well.

**Comment by vikas5914** [comment]
@stojankukrika You can download from here: https://packages.macports.org/php82-mongodb/

**Reply by stojankukrika** [reply]
Which one and what to do with it?

**Reply by vikas5914** [reply]
Read the file name. It has os name and CPU type. Download the zip folder, and inside that zip folder, you will get the `.so` file, which is already a compiled extension. You can copy that to the heard folder and use `php.ini` to point it.

**Reply by stojankukrika** [reply]
Thank you vikas5914 on your time to explain to me! Have a nice day friend!

**Comment by chrisbaswell** [comment]
Snowflake PDO - https://github.com/snowflakedb/pdo_snowflake

**Comment by StefanNeuser** [comment]
mcrypt for php 7.4 to support Laravel 4.2 :-)

**Comment by iambateman** [comment]
gettext. 

I _think_ switching to herd made something like `__('Translatable')` break, and as far as I can tell, there's no way to enable it within Herd.

**Comment by nam-co** [comment]
I haven't found any issues with json translations on any project, maybe is something else

**Comment by thijsdewitt** [comment]
`ext-xsl` Would be nice for us Magento developers

**Reply by StanBarrows** [reply]
This would be awesome to use https://github.com/saloonphp/xml-wrangler

**Comment by milhouse1337** [comment]
I've compiled a list of extensions that were included in the brew build but are absent in this one. Here's the breakdown, along with my reasoning for their potential inclusion:

- `gettext` This is widely used for translation purposes and is not available via PECL.
- `ldap` Not available on PECL.
- `odbc` Available on PECL but not maintained anymore.
- `pdo_dblib` Available on PECL but not maintained anymore.
- `pdo_odbc` Available on PECL but not maintained anymore.
- `pspell` Not available on PECL.
- `random` Not available on PECL.
- `tidy` Available on PECL but not maintained anymore.
- `xsl` This is used for XSL transformations and is not available via PECL.

In summary, I believe we should at least include `gettext`, `ldap`, `odbc`, `pdo_dblib`, `pdo_odbc`, `pspell`, `random` and `xsl`, given that there aren't alternative installation methods for them via PECL.

EDIT: Missed a few modules, sorry. If you guys want to list the available modules in your build there is this handy command: `php -m` 😅

**Comment by pyrou** [comment]
Pecl still requires to enable (manually or not) extension in php.ini and problem is: the php.ini file is shared across all php versions in herd.

**Comment by jason-beach** [comment]
@pyrou ,

With Herd's latest build, there are now php.ini files for each version of PHP. So you can add extensions per version of PHP.

**Comment by martinemmert** [comment]
Can we pay or support you otherwise to include the proposal of @milhouse1337 ? 

I am currently stuck building an MVP and need XSL badly.
Tinker around and lose more time with `phpize` is something I want to avoid.

You did an awesome job and provided us Herd for free.
Saving me a ton of time is worth my money 👍 

edit:

@vikas5914 pointed me in the right direction a few comments further below

**Comment by pryley** [comment]
@martinemmert Why not just use Laravel Valet?

**Comment by martinemmert** [comment]
@pryley never used it before and Herd was as easy as it can get setting up a php environment. 

It just works and I don't have to care about it that much. 

Nevertheless, I'll take a look at Valet. 

Thanks

**Comment by hans0801** [comment]
> @pyrou ,
> 
> With Herd's latest build, there are now php.ini files for each version of PHP. So you can add extensions per version of PHP.

Hi @jason-beach , 
Thanks fo the information. 
So the new released version had every ini file right? 
Actually i need ldap extension on this herd, could you help me for add those extension when i will use herd? 
Thanks

**Comment by vikas5914** [comment]
@martinemmert : Go to this link: https://packages.macports.org/php82-xsl/ (you can php version) and download the version you want (arm64 for M1 mac), and it will give the `.so` file that you can directly put in herd directory and change the php.ini file.

**Comment by esseremmerik** [comment]
The solution by @vikas5914 is perfect! Don't forget to open the extension file in the finder with a random application and allow the downloaded file from the internet (unknown supplier), otherwise PHP is not able to load the extension (if you have enabled strict file restrictions in OSX)

**Comment by baskinvolkan** [comment]
> `sqlsrv` 
> 
> SQL Server is a database that has first-party support through Laravel.

🙏

**Comment by harizinside** [comment]
It would be awesome to add Mongodb

**Reply by iprastha** [reply]
Managed to get it going in Herd, by installing mongodb through PECL
```
sudo pecl install mongodb
```
Make sure to note down the final output where the **mongodb.so** is located, and use that path in
the following guide to add extension manually to Herd php.ini config
https://herd.laravel.com/docs/1/advanced-usage/additional-extensions

**Comment by rrrrando** [comment]
Add `imagick-extra` for svg support. Right now can't use spatie media library with svg files.

**Comment by gitekDev** [comment]
Add IMAP :=)

**Comment by StanBarrows** [comment]
The XSL extension for using https://github.com/saloonphp/xml-wrangler

**Comment by ryanmortier** [comment]
I see in the documentation the ability to add PHP extensions: https://herd.laravel.com/docs/1/advanced-usage/additional-extensions

Does this solve this thread? I'm reformatting my macbook right now and wondering if I need to bother with Valet now.

**Reply by nicekiwi** [reply]
Not for all extensions. Some like `gettext` need to be enabled at compile time and cannot be added afterwards. E.g. the Homebrew version of PHP includes such extensions by default.

**Comment by s7thamon** [comment]
ldap prevents me from using herd at the moment as this is not available in pecl

**Comment by ericlbarnes** [comment]
I just needed to install Imap, and here is how I did it, which seems to work: 

```
brew install php
```

```
brew install php-imap
```

Then, in Herds php.ini (Library/Application Support/Herd/config/php/83/php.ini):

```
extension=/opt/homebrew/lib/php/20230831/imap.so
```

I used the [docs](https://herd.laravel.com/docs/1/advanced-usage/additional-extensions) as a guide. Hope this helps anyone else.

**Reply by mshannaq** [reply]
@ericlbarnes what if I want to install php gettext ext because Herd comes without ext-gettext support

while I run
```
brew install brew install gettext
```

I got `No available formula with the name "php-gettext"`

**Comment by geowrgetudor** [comment]
`ext-xsl` would be nice. https://github.com/veewee/xml requires it

**Comment by mpociot** [comment]
New PHP builds are currently distributed (a new 8.3 version for ARM64 is already available).
These new builds include a couple of extensions by default:

* imap
* ldap
* xsl

🎉 

If you have previously installed one of those extensions manually, you might need to remove the custom extension.

**Reply by geowrgetudor** [reply]
Thanks!

**Reply by aurawindsurfing** [reply]
how about ghostsctript and imagicks?

I understand oci8 is rare but those are 2 very common extensions.

Thanks @mpociot

**Comment by mshannaq** [comment]
gmp

**Comment by dasundev** [comment]
`ssh2` also needed.

**Comment by cyppe** [comment]
I have separate thread about it, but adding it here just in case this is where the focus is from Herd devs.

But either add serializers and compression to the compiled in phpredis extention.  For my specific use case it's LZ4 and Igbinary that is required. But if it should be added to compiled I think it makes sense to add a few more. As least same as Forge:

forge@ma-dev-04:~$ php --ri redis

redis

Redis Support => enabled
Redis Version => 6.0.2
Redis Sentinel Version => 1.0
**Available serializers => php, json, igbinary**
**Available compression => lzf, zstd, lz4**


---------------------------------------


**Alternative approach**

If that is not interesting for you for some reason, I would like that you stop compiling the redis extention and instead load it as extention so I can disable your redis.so and load my own. 

My full thread here:

https://github.com/beyondcode/herd-community/discussions/470

**Comment by RyanRoberts** [comment]
Avif file format in Imagemagick would be appreciated. #935

**Comment by lucarossi77** [comment]
The pdo-firebird extension is the missing link for my projects to fully transition to herd

**Comment by Lutacon** [comment]
odbc & pdo_odbc aren't available to install with pecl

It would be nice to have them in Herd.

**Comment by helderneves91** [comment]
Does anyone work with ODBC? Can't get the connection to work since pecl can't install the extension.

**Reply by Lutacon** [reply]
I had to switch back to using docker(sail). Compile manually the extension. If you want i can send you the Dockerfile

**Comment by vikas5914** [comment]
@helderneves91 Download the driver from here: https://learn.microsoft.com/en-us/sql/connect/php/download-drivers-php-sql-server?view=sql-server-ver16

**Comment by brunogoossens** [comment]
I have installed herd-lite (php.new for linux) and i want to install extra extensions. I tried the new Pie extension installer but i noticed phpize (php-dev) is not installed on my machine. Maybe shipping phpize with herd-lite is an option so users can install extra extensions?
https://github.com/php/pie

**Reply by codedusting** [reply]
@brunogoossens found any solution here? On how to install new extension apart from what comes with this php.new script?

**Reply by brunogoossens** [reply]
Not using herd-lite

**Comment by kasperhartwich** [comment]
How about `ext-uuid`?

It is supported if we look [here](https://static-php.dev/en/guide/extensions.html).

**Comment by SecondeJK** [comment]
Re: comments about PIE.
Request to support PIE and include it in the documentation, so fetching new extensions will be pretty painless.

**Comment by LukaszJaro** [comment]
Lightweight Xdebug by default?

https://laravel-news.com/php-debugger-a-lightweight-xdebug-alternative-built-for-speed
